File: app/src/main/java/com/winlator/steampipeserver/SteamPipeServer.java
package com.winlator.steampipeserver;

import android.util.Log;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;

public class SteamPipeServer {
    private static final int PORT = 34865;
    private ServerSocket serverSocket;
    private boolean running;

    private int readNetworkInt(DataInputStream input) throws IOException {
        return Integer.reverseBytes(input.readInt());
    }

    private void writeNetworkInt(DataOutputStream output, int value) throws IOException {
        output.writeInt(Integer.reverseBytes(value));
    }

    public void start() {
        running = true;
        new Thread(() -> {
            try {
                serverSocket = new ServerSocket();
                serverSocket.setReuseAddress(true);
                serverSocket.bind(new InetSocketAddress(PORT));
                Log.d("SteamPipeServer", "Server started on port " + PORT);

                while (running) {
                    try {
                        Socket clientSocket = serverSocket.accept();
                        handleClient(clientSocket);
                    } catch (IOException e) {
                        if (running) {
                            Log.e("SteamPipeServer", "Error accepting client connection", e);
                        } else {
                            Log.d("SteamPipeServer", "Server socket closed during shutdown");
                        }
                        break;
                    }
                }
            } catch (IOException e) {
                Log.e("SteamPipeServer", "Server error", e);
            } finally {
                Log.d("SteamPipeServer", "Server thread exiting");
            }
        }).start();
    }

    private void handleClient(Socket clientSocket) {
        new Thread(() -> {
            try {
                DataInputStream input = new DataInputStream(
                        new BufferedInputStream(clientSocket.getInputStream()));
                DataOutputStream output = new DataOutputStream(
                        new BufferedOutputStream(clientSocket.getOutputStream()));

                while (running && !clientSocket.isClosed()) {
                    try {
                        if (input.available() > 0) {
                            int messageType = readNetworkInt(input);

                            switch (messageType) {
                                case RequestCodes.MSG_INIT:
                                    Log.d("SteamPipeServer", "Received MSG_INIT");
                                    writeNetworkInt(output, 1);
                                    output.flush();
                                    break;
                                case RequestCodes.MSG_SHUTDOWN:
                                    Log.d("SteamPipeServer", "Received MSG_SHUTDOWN");
                                    clientSocket.close();
                                    break;
                                case RequestCodes.MSG_RESTART_APP:
                                    Log.d("SteamPipeServer", "Received MSG_RESTART_APP");
                                    int appId = input.readInt();
                                    writeNetworkInt(output, 0); // Send restart not needed
                                    break;
                                case RequestCodes.MSG_IS_RUNNING:
                                    Log.d("SteamPipeServer", "Received MSG_IS_RUNNING");
                                    writeNetworkInt(output, 1); // Send Steam running status
                                    break;
                                case RequestCodes.MSG_REGISTER_CALLBACK:
                                    Log.d("SteamPipeServer", "Received MSG_REGISTER_CALLBACK");
                                    break;
                                case RequestCodes.MSG_UNREGISTER_CALLBACK:
                                    Log.d("SteamPipeServer", "Received MSG_UNREGISTER_CALLBACK");
                                    break;
                                case RequestCodes.MSG_RUN_CALLBACKS:
                                    Log.d("SteamPipeServer", "Received MSG_RUN_CALLBACKS");
                                    break;
                                default:
                                    Log.w("SteamPipeServer", "Unknown message type: " + messageType);
                                    break;
                            }
                        }

                        Thread.sleep(10);
                    } catch (InterruptedException e) {
                        Log.d("SteamPipeServer", "Client thread interrupted");
                        break;
                    } catch (IOException e) {
                        if (running) {
                            Log.e("SteamPipeServer", "Error reading from client", e);
                        }
                        break;
                    }
                }
            } catch (IOException e) {
                Log.e("SteamPipeServer", "Client handler error", e);
            } finally {
                try {
                    clientSocket.close();
                } catch (IOException e) {
                    Log.e("SteamPipeServer", "Error closing client socket", e);
                }
                Log.d("SteamPipeServer", "Client thread exiting");
            }
        }).start();
    }

    public void stop() {
        running = false;
        try {
            if (serverSocket != null) {
                serverSocket.close();
            }
        } catch (IOException e) {
            Log.e("SteamPipeServer",e.getLocalizedMessage());
        }
    }
}

File: winlator/src/main/java/com/winlator/steampipeserver/SteamPipeServer.kt
package com.winlator.steampipeserver

import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import timber.log.Timber

class SteamPipeServer {

    companion object {
        private const val PORT = 34865
    }

    private var serverSocket: ServerSocket? = null

    private var running = false

    @Throws(IOException::class)
    private fun readNetworkInt(input: DataInputStream): Int {
        return Integer.reverseBytes(input.readInt())
    }

    @Throws(IOException::class)
    private fun writeNetworkInt(output: DataOutputStream, value: Int) {
        output.writeInt(Integer.reverseBytes(value))
    }

    fun start() {
        running = true
        Thread(
            Runnable {
                try {
                    serverSocket = ServerSocket()
                    serverSocket!!.setReuseAddress(true)
                    serverSocket!!.bind(InetSocketAddress(PORT))
                    Timber.d("Server started on port %s", PORT)

                    while (running) {
                        val clientSocket: Socket = serverSocket!!.accept()
                        // clientSocket.setTcpNoDelay(true);
                        // clientSocket.setSoTimeout(5000);  // 5 second timeout
                        handleClient(clientSocket)
                    }
                } catch (e: IOException) {
                    Timber.e(e, "Server error")
                }
            },
        ).start()
    }

    private fun handleClient(clientSocket: Socket) {
        Thread(
            Runnable {
                try {
                    val input = DataInputStream(BufferedInputStream(clientSocket.getInputStream()))
                    val output = DataOutputStream(BufferedOutputStream(clientSocket.getOutputStream()))

                    while (running && !clientSocket.isClosed) {
                        if (input.available() > 0) {
                            val messageType = readNetworkInt(input)

                            // Log.d("SteamPipeServer", "Received message: " + messageType);
                            when (messageType) {
                                RequestCodes.MSG_INIT -> {
                                    Timber.d("Received MSG_INIT")
                                    writeNetworkInt(output, 1)
                                    output.flush()
                                }

                                RequestCodes.MSG_SHUTDOWN -> {
                                    Timber.d("Received MSG_SHUTDOWN")
                                    clientSocket.close()
                                }

                                RequestCodes.MSG_RESTART_APP -> {
                                    Timber.d("Received MSG_RESTART_APP")
                                    // val appId = input.readInt()
                                    writeNetworkInt(output, 0) // Send restart not needed
                                }

                                RequestCodes.MSG_IS_RUNNING -> {
                                    Timber.d("Received MSG_IS_RUNNING")
                                    writeNetworkInt(output, 1) // Send Steam running status
                                }

                                RequestCodes.MSG_REGISTER_CALLBACK -> Timber.d("Received MSG_REGISTER_CALLBACK")

                                RequestCodes.MSG_UNREGISTER_CALLBACK -> Timber.d("Received MSG_UNREGISTER_CALLBACK")

                                RequestCodes.MSG_RUN_CALLBACKS -> Timber.d("Received MSG_RUN_CALLBACKS")

                                else -> Timber.w("Unknown message type: %s", messageType)
                            }
                        }
                    }
                } catch (e: IOException) {
                    Timber.e(e, "Client handler error")
                }
            },
        ).start()
    }

    fun stop() {
        running = false

        try {
            serverSocket?.close()
        } catch (e: IOException) {
            Timber.w(e)
        }
    }
}
